### PR TITLE
Handle changes related to new API (Serverless spec)

### PR DESCRIPTION
**Pull request by vdayanand**
ref: https://docs.pinecone.io/guides/operations/migrate-to-the-new-api
- adds a new objects with suffix "v3" to handle the new API

## Summary by CodeRabbit

- **New Features**
	- Introduced new functions for HTTP POST and DELETE requests to enhance data manipulation capabilities.
	- Added a new file defining an interface for the Pinecone v3 API, including structures and functions for managing vector indices.

- **Improvements**
	- Updated display formats for `PineconeIndex` and `PineconeVector` for clearer output.

- **Tests**
	- Added a new test suite for the Pinecone v3 API to validate core functionalities, including index creation and vector operations.

### CODE REVIEW

File: src/v3.jl [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+
+# Example
+```julia
+Pinecone.delete_index(context, Pinecone.Index("index-to-delete"))
+```
+"""
+function delete_index(ctx::PineconeContextv3, indexname::String)
+    url = pineconeMakeURLForIndex(indexname)
+    response = pineconeHTTPDelete(url, ctx.apikey)
+end
```
[the remark this answers is not in the document]
Comment (vdayanand) [reply src/v3.jl:168]: I have kept it consistent with how its handled in the existing code

File: src/v3.jl [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+"""
+function delete(ctx::PineconeContextv3, indexobj::PineconeIndexv3, ids::Array{String}, deleteall::Bool=false, namespace::String="")
+    if(length(ids) > MAX_DELETE)
+        throw(ArgumentError("Max number of vectors per delete is " * string(MAX_DELETE)))
+    end
+    url = pineconeMakeURLForIndex(indexobj, ENDPOINTDELETE)
+    postbody = Dict{String, Any}("ids"=>ids, "namespace"=>namespace, "deleteAll"=>deleteall)
+    response = pineconeHTTPPost(url, ctx.apikey, JSON3.write(postbody))
+    return String(response.body)
+end
```
[the remark this answers is not in the document]
Comment (vdayanand) [reply src/v3.jl:221]: I have kept it consistent with how its handled in the existing code

File: src/v3.jl [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+    end
+    url = pineconeMakeURLForIndex(indexobj, ENDPOINTQUERYINDEX)
+    body = Dict{String, Any}("topK"=>topk, "vector"=>vector, "includeValues"=>includevalues, "includeMetadata"=>includemeta, "namespace"=>namespace)
+    if(length(filter) > 0)
+        body["filter"] = filter;
+    end
+    postbody = JSON3.write(body)
+    response = pineconeHTTPPost(url, ctx.apikey, postbody)
+    return String(response.body)
+end #query
```
[the remark this answers is not in the document]
Comment (vdayanand) [reply src/v3.jl:317]: I have kept it consistent with how its handled in the existing code

File: src/v3.jl [hunk 4]
```diff
@@ -0,0 +1,317 @@
...
+    response = pineconeHTTPPost(url, ctx.apikey, JSON3.write(body))
+end
+
+function list_index_objects(ctx::PineconeContextv3)
+    response = pineconeHTTPGet("https://api.pinecone.io/indexes", ctx.apikey)
+    ## do better here
+    indexes = JSON3.read(String(response.body), Dict{Symbol, Any})[:indexes]
+    a = JSON3.write(indexes)
+    JSON3.read(JSON3.write(indexes), Vector{PineconeIndexv3})
+end
```
[the remark this answers is not in the document]
Comment (vdayanand) [reply src/v3.jl:110]: This suggestion doesn't work

[the remark this answers is not in the document]
Comment (vdayanand) [reply src/v3.jl]: ok applied the change

File: src/v3.jl [hunk 5]
```diff
@@ -0,0 +1,317 @@
...
+julia> Pinecone.Index("my-index")
+PineconeIndex connected to my-index
+```
+"""
+
+function Index(ctx::PineconeContextv3, indexname::String)
+    url = pineconeMakeURLForIndex(indexname)
+    response = pineconeHTTPGet(url, ctx.apikey)
+    return JSON3.read(String(response.body), PineconeIndexv3)
+end
```
[the remark this answers is not in the document]
Comment (vdayanand) [reply src/v3.jl:143]: I have kept it consistent with how its handled in the existing code

File: src/v3.jl [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+
+function describe_index_stats(ctx::PineconeContextv3, indexobj::PineconeIndexv3)
+    url = pineconeMakeURLForIndex(indexobj, ENDPOINTDESCRIBEINDEXSTATS)
+    response = pineconeHTTPGet(url, ctx.apikey)
+    return String(response.body)
+end
```
[the remark this answers is not in the document]
Comment (vdayanand) [reply src/v3.jl:149]: I have kept it consistent with how its handled in the existing code

File: src/v3.jl [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+    end
+    url = pineconeMakeURLForIndex(indexobj, ENDPOINTUPSERT)
+    body = Dict{String, Any}("vectors" => vectors)
+    if namespace !== nothing && namespace != ""
+        body["namespace"] = namespace
+    end
+    postbody = JSON3.write(body)
+    response = pineconeHTTPPost(url, ctx.apikey, postbody)
+    return String(response.body)
+end #upsert
```
[the remark this answers is not in the document]
Comment (vdayanand) [reply src/v3.jl:195]: I have kept it consistent with how its handled in the existing code

File: src/v3.jl [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+                "spec"=>Dict(
+                    "serverless"=>Dict(
+                        "cloud"=>cloud,
+                        "region"=>region
+                    )
+                )
+                )
+    url = pineconeMakeURLForIndex()
+    response = pineconeHTTPPost(url, ctx.apikey, JSON3.write(body))
+end
```
[the remark this answers is not in the document]
Comment (vdayanand) [reply src/v3.jl:102]: I have kept it consistent with how its handled in the existing code

File: src/v3.jl [hunk 9]
```diff
@@ -0,0 +1,317 @@
...
+function fetch(ctx::PineconeContextv3, indexobj::PineconeIndexv3, ids::Array{String}, namespace::String="")
+    if(length(ids) > MAX_FETCH)
+        throw(ArgumentError("Max number of vectors per fetch is " * string(MAX_FETCH)))
+    end
+    renamedids = ["ids=$row" for row in ids]
+    urlargs = "?" * join(renamedids, "&") * "&namespace=$namespace"
+    url = pineconeMakeURLForIndex(indexobj, ENDPOINTFETCH) * urlargs
+    response = pineconeHTTPGet(url, ctx.apikey)
+    return String(response.body)
+end
```
[the remark this answers is not in the document]
Comment (vdayanand) [reply src/v3.jl:245]: I have kept it consistent with how its handled in the existing code

**Comment by vdayanand** [comment]
@tullytim could you please review the changes ?

**Comment by vdayanand** [comment]
@tullytim I have fixed the CI failure. Could you retrigger it again?

**Comment by vdayanand** [comment]
The current failure seems to related to invalid API key in CI as it works fine locally?
```
 """
HTTP/1.1 401 Unauthorized
x-pinecone-api-version: 2024-04
X-Cloud-Trace-Context: 6168af481f6840608399c5de39fa395a
Date: Sat, 10 Aug 2024 22:34:23 GMT
Content-Type: text/html
Server: Google Frontend
Content-Length: 15
Via: 1.1 google
Alt-Svc: h3=":443"; ma=2592000,h3-29=":443"; ma=2592000

Invalid API Key""")```

**Comment by tullytim** [comment]
I'll rerun it with a new api key

**Pull request merged**